src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
+    WaitChildExit(false);
+}
+
+void SubprocessSet::WaitChildExit(bool block) {
+  int status = 0;
+  int options = block ? 0 : WNOHANG;
+  pid_t pid = waitpid(-1, &status, options);
+  if (pid == -1)
+    Fatal("waitpid: %s", strerror(errno));
+  if (pid == 0)
+    return;
+  for (vector<Subprocess*>::iterator i = running_.begin();
```

Comment:
This loops kinda duplicates code in DoWork() (the one right after 'nfds_t cur_nfd = 0;')
Wouldn't it be simpler to not have this additional loop here at all, rather to call HandleChildExit() just before the return from DoWork(), after all pipes are drained?